File: vanguard_os_legacy/shell.py
import glob
import os
import sys

from core_ai import VanguardAI  # Переносим ИИ под управление Шелла
from executor import VanguardExecutor  # Импортируем наш созданный запуск процессов


class VanguardShell:
    def __init__(self):
        # Подключаем наш исполнитель скриптов
        self.executor = VanguardExecutor()
        self.ai_core = VanguardAI()

    def handle_command(self, query):
        """Единая точка обработки ВСЕХ команд и запросов Vanguard OS"""
        cmd = query.strip().lower()

        # 1. КОМАНДА ВЫХОДА ИЗ ОС (/exit или /shutdown)
        if cmd in ["/exit", "/shutdown", "выход"]:
            # Возвращаем специальный флаг завершения работы
            return ">> SHUTTING DOWN VANGUARD ENVIRONMENT...", "SYSTEM_EXIT"

        # 2. КОМАНДА ОЧИСТКИ ПАПКИ (/wipe)
        elif cmd == "/wipe":
            log = ">> INITIATING STORAGE PURGE PROTOCOL...\n"
            generated_files = glob.glob("vanguard_gen_*.py")
            deleted_count = 0
            for file_path in generated_files:
                try:
                    os.remove(file_path)
                    log += f">> PURGED: {file_path}\n"
                    deleted_count += 1
                except Exception as ex:
                    log += f">> ERROR DELETING {file_path}: {ex}\n"
            log += f"\n>> PURGE COMPLETE. TOTAL FILES DESTROYED: {deleted_count}\n"
            return log, "DISPLAY_OUTPUT"


        # 3. КОМАНДА ПОЛНОГО СБРОСА ЭКРАНА (/reset)
        elif cmd == "/reset":
            # [ДОБАВИТЬ ЭТУ СТРОКУ] Намертво вычищаем контекст из нейросети
            self.ai_core.clear_memory()

            log = (
                "==================================================\n"
                "                VANGUARD SYSTEM RESET             \n"
                "==================================================\n"
                ">> WIPING DISPLAY BUFFER...\n"
                ">> DISCONNECTING PREVIOUS NEURAL CONTEXT... DONE\n"
                ">> VANGUARD KERNEL: READY FOR NEW INSTRUCTIONS.\n"
            )
            return log, "DISPLAY_OUTPUT"

        # 4. КОМАНДА АСИНХРОННОГО ЗАПУСКА СКРИПТОВ (/run <id>)
        elif cmd.startswith("/run"):
            parts = query.split()
            if len(parts) < 2:
                return (
                    ">> ERROR: Specify file ID. Example: /run 553\n",
                    "DISPLAY_OUTPUT",
                )

            file_id = parts

            def async_execute():
                execution_log = self.executor.execute_script(file_id)
                print(f"\n{execution_log}\n")

            import threading

            threading.Thread(target=async_execute, daemon=True).start()
            return (
                f">> INITIATING PROCESS FOR VANGUARD_GEN_{file_id}.PY...\n>> STREAMING OUTPUT TO PYCHARM CONSOLE.\n",
                "DISPLAY_OUTPUT",
            )
            # [NEW] КОМАНДА: ВЫВОД СПИСКА СГЕНЕРИРОВАННЫХ СКРИПТОВ (/apps)
        elif cmd == "/apps":
            generated_files = glob.glob("vanguard_gen_*.py")
            if not generated_files:
                return (
                    ">> VANGUARD REGISTRY: No generated applications found.\n",
                    "DISPLAY_OUTPUT",
                )

            log = (
                "==================================================\n"
                "             VANGUARD APPLICATION REGISTRY        \n"
                "==================================================\n"
            )
            for file_path in generated_files:
                # Получаем размер файла в КБ
                file_size = os.path.getsize(file_path) / 1024
                log += f"-> [ID: {file_path[13:-3]}] | FILE: {file_path} | SIZE: {file_size:.2f} KB\n"
            log += (
                "==================================================\n"
                ">> TYPE: /run <id> to execute targeted script.\n"
            )
            return log, "DISPLAY_OUTPUT"
            # [NEW] КОМАНДА: ДЕТАЛЬНЫЙ АУДИТ ЖЕЛЕЗА ПК (/hardware)
        elif cmd == "/hardware":
            import psutil

            # Сбор информации о памяти и ядрах
            cores_physical = psutil.cpu_count(logical=False)
            cores_logical = psutil.cpu_count(logical=True)
            ram = psutil.virtual_memory()

            log = (
                "==================================================\n"
                "               VANGUARD HARDWARE AUDIT            \n"
                "==================================================\n"
                ">> ARCHITECTURE: AMD64/x86_64 SECURED ENVIRONMENT\n"
                f">> CPU INFRASTRUCTURE: {cores_physical} Physical Cores | {cores_logical} Threads\n"
                f">> SYSTEM RAM CAPACITY: {ram.total / (1024**3):.2f} GB Total\n"
                f">> RAM CURRENTLY AVAILABLE: {ram.available / (1024**3):.2f} GB Free\n"
                f">> PLATFORM INTEGRITY: 100% SUVEREIGN KERNEL\n"
                "==================================================\n"
            )
            return log, "DISPLAY_OUTPUT"
            # [NEW] КОМАНДА: ВЫВОД РУКОВОДСТВА ПО СИСТЕМЕ (/help)
        elif cmd == "/help":
            log = (
                "==================================================\n"
                "               VANGUARD SHELL COMMANDS            \n"
                "==================================================\n"
                ">> /help     -> Display this tactical instruction log.\n"
                ">> /apps     -> List all AI auto-compiled python applications.\n"
                ">> /run <id> -> Execute targeted python script in background thread.\n"
                ">> /hardware -> Perform deep hardware audit and scan resources.\n"
                ">> /wipe     -> Purge and destroy all compiled vanguard scripts.\n"
                ">> /reset    -> Clear display buffer and reset neural channel.\n"
                ">> /exit     -> Soft shutdown of the Vanguard OS infrastructure.\n"
                "==================================================\n"
            )
            return log, "DISPLAY_OUTPUT"

        # 5. ЕСЛИ ЭТО ОБЫЧНЫЙ ЗАПРОС К ИИ — ПЕРЕДАЕМ УПРАВЛЕНИЕ СТРИМЕРУ
        else:
            # Возвращаем пустую строку и флаг, разрешающий потоковый ИИ
            return "", "AI_STREAM_ALLOWED"

File: vanguard_os_rust/src/shell.rs
use std::fs::{self, File};
use std::io::Write;
use std::thread;
use glob::glob;
use regex::Regex;
use rand::Rng;
use sysinfo::System;

use crate::executor::VanguardExecutor;

pub struct VanguardAI {
    pub history: String,
}

impl VanguardAI {
    pub fn new() -> Self {
        Self { history: String::from("System: Инициализирован встроенный ИИ-терминал.\n") }
    }
    pub fn clear_memory(&mut self) {
        self.history = String::from("System: Инициализирован встроенный ИИ-терминал.\n");
    }
    pub fn save_code_if_exists(&mut self, full_text: &str) -> Option<String> {
        self.history.push_str(&format!("{}\n", full_text));
        let re = Regex::new(r"(?s)```python\n(.*?)\n```").unwrap();

        if let Some(captures) = re.captures(full_text) {
            if let Some(code_block) = captures.get(1) {
                let gen_id = rand::thread_rng().gen_range(100..1000);
                let filename = format!("vanguard_gen_{}.py", gen_id);

                if let Ok(mut file) = File::create(&filename) {
                    if file.write_all(code_block.as_str().as_bytes()).is_ok() {
                        return Some(filename);
                    }
                }
            }
        }
        None
    }
}

#[derive(Debug, PartialEq)]
pub enum ShellAction {
    SystemExit,
    DisplayOutput(String),
    AiStreamAllowed,
}

pub struct VanguardShell {
    pub ai_core: VanguardAI,
}

impl VanguardShell {
    pub fn new() -> Self {
        Self { ai_core: VanguardAI::new() }
    }
    pub fn handle_command(&mut self, query: &str) -> ShellAction {
        let trimmed = query.trim();
        let cmd_lower = trimmed.to_lowercase();
        let parts: Vec<&str> = trimmed.split_whitespace().collect();

        if parts.is_empty() {
            return ShellAction::DisplayOutput(String::new());
        }

        match parts[0].to_lowercase().as_str() {
            "/exit" | "/shutdown" | "выход" => ShellAction::SystemExit,
            "/wipe" => {
                let mut log = String::from(">> INITIATING STORAGE PURGE PROTOCOL...\n");
                let mut deleted_count = 0;
                if let Ok(entries) = glob("vanguard_gen_*.py") {
                    for entry in entries.flatten() {
                        if fs::remove_file(&entry).is_ok() {
                            log.push_str(&format!(">> PURGED: {}\n", entry.display()));
                            deleted_count += 1;
                        }
                    }
                }
                log.push_str(&format!("\n>> PURGE COMPLETE. TOTAL FILES DESTROYED: {}\n", deleted_count));
                ShellAction::DisplayOutput(log)
            }
            "/reset" => {
                self.ai_core.clear_memory();
                ShellAction::DisplayOutput(String::from(">> WIPING DISPLAY BUFFER...\n>> PREVIOUS NEURAL CONTEXT DROPPED.\n"))
            }
            "/run" => {
                if parts.len() < 2 {
                    return ShellAction::DisplayOutput(">> ERROR: Specify file ID. Example: /run 553\n".to_string());
                }
                let file_id = parts[1..].join(" ");
                let file_id_clone = file_id.clone();
                thread::spawn(move || {
                    let execution_log = VanguardExecutor::execute_script(&file_id_clone);
                    println!("\n{}\n", execution_log);
                });
                ShellAction::DisplayOutput(format!(">> INITIATING BACKGROUND TASK FOR APP ID: {}...\n", file_id))
            }
            "/apps" => {
                let mut log = "=== VANGUARD APPLICATION REGISTRY ===\n".to_string();
                let mut found = false;
                if let Ok(entries) = glob("vanguard_gen_*.py") {
                    for entry in entries.flatten() {
                        found = true;
                        let file_size = fs::metadata(&entry).map(|m| m.len() as f64 / 1024.0).unwrap_or(0.0);
                        let file_name = entry.to_string_lossy();
                        let id = if file_name.len() > 16 { &file_name[13..file_name.len() - 3] } else { "???" };
                        log.push_str(&format!("-> [ID: {}] | FILE: {} | SIZE: {:.2} KB\n", id, file_name, file_size));
                    }
                }
                if !found { return ShellAction::DisplayOutput(">> REGISTRY EMPTY.\n".to_string()); }
                ShellAction::DisplayOutput(log)
            }
            "/hardware" => {
                let mut sys = System::new_all();
                sys.refresh_cpu_usage();
                sys.refresh_memory();
                ShellAction::DisplayOutput(format!(">> CPU THREADS: {}\n>> RAM CAPACITY: {:.2} GB Total\n", sys.cpus().len(), sys.total_memory() as f64 / 1024.0 / 1024.0 / 1024.0))
            }
            "/help" => ShellAction::DisplayOutput(String::from(">> COMMANDS: /help, /apps, /run <id>, /hardware, /wipe, /reset, /exit\n")),
            _ => {
                if cmd_lower.starts_with('/') {
                    ShellAction::DisplayOutput(format!(">> vsh: command not found: {}\n", cmd_lower))
                } else {
                    ShellAction::AiStreamAllowed
                }
            }
        }
    }
}